### ISSUE TITLE
Add support to upgrade Orchestrator Version for Spot Node Pools

### ISSUE THREAD

**Issue by thejosephstevens** [Azure/AKS#2134]
**What happened**:
We're attempting to upgrade clusters to 1.19 from 1.18 with Terraform. On attempting to upgrade, the node pool refused to upgrade because it's a spot pool, stating that it can't guarantee uptime.
**What you expected to happen**:
I expected AKS to roll nodes with a best effort for maintaining uptime with no guarantee of full uptime.
**How to reproduce it (as minimally and precisely as possible)**:
Create an aks cluster with a non-default spot node pool of version 1.18.*, change the version of both to 1.19.7 in terraform, and apply.

**Anything else we need to know?**:
Because of the current design with this block on upgrade functionality, we can't upgrade clusters during business hours. Brief interruptions during a roll that quickly recover are acceptable, but with the current design we have no option but to incur very heavy downtime by deleting the pool and recreating, or build our own rollover mechanism in terraform in order to operate our clusters, both of which seem like a waste of time for me to build when I'm paying for a managed k8s service. Even more specifically, this isn't an issue I have with EKS or GKE (or frankly even kops), they both either have mechanisms to upgrade node pools in the background (regardless of purchase option), or to do orchestrated rolls (again, regardless of VM purchase option).
**Environment**:
- Kubernetes version (use `kubectl version`): 1.18.*
- Size of cluster (how many worker nodes are in the cluster?): 1-30
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.): microservices, spark jobs
- Others:

**Comment by palma21** [comment]
Hi @thejosephstevens as per https://docs.microsoft.com/en-us/azure/aks/spot-node-pool#limitations
> You can't upgrade a spot node pool since spot node pools can't guarantee cordon and drain. You must replace your existing spot node pool with a new one to do operations such as upgrading the Kubernetes version. To replace a spot node pool, create a new spot node pool with a different version of Kubernetes, wait until its status is Ready, then remove the old node pool.

The reason is by doing this you can actually have a much more reliable upgrade process so just upgrade your control plane, create a new spot nodepool with your desired version and drain/delete the old spot pool.

Would this work for you?

**Comment by emaildanwilson** [comment]
I'm not the OP but IMO this doesn't work because there isn't a way to make the workflow you're suggesting automated with a terraform apply. If the terraform provider did this automatically then it would be fine. Honestly, I don't even see why we need to specify a version for each spot node pool. What is the use case where someone wouldn't want the next spot vm to match the cluster version automatically?

**Comment by thejosephstevens** [comment]
Hey @palma21 we did that, and it's turned our terraform into a mess because we need to keep track of whether we're on an `a` or `b` pool (which continually has to be committed and updated in source). It effectively means we lose the advantage of terraform because it's now impossible to describe a stable spec for our clusters. Rather than just declaring a spec and the provider taking care of the details, we now need to have a state machine in our heads that we're operating around in order to perform basic cluster operations because there's a whole set of invalid state transitions from the provider.

The problem I see here is that we chose spot nodes because they were a good value proposition for our fault-tolerant workloads, and we set up configuration in kubernetes so that it orchestrates around disruptions (like the disruptions you would see in a node pool role). This has worked great for us and given us a lot of flexibility to move quickly and deliver value to our customers, however this block on upgrading spot pools comes exclusively from AKS (again, not an issue on other clouds or with other k8s providers), and forces us to work around it as an arbitrary limitation intended to keep us from running into scenarios that pose no threat to us.

So yes, we are currently playing ping pong between spot pools, but no it's not a satisfactory workaround.

**Comment by emaildanwilson** [comment]
This issue should not be closed.

**Comment by thejosephstevens** [comment]
@palma21 is there going to be any action taken on this? Being a bit blunt, the limitation of keeping us from performing in-place node pool upgrades creates a problem where there wasn't any before. If the block was removed and the link to https://docs.microsoft.com/en-us/azure/aks/spot-node-pool#limitations just said "AKS makes no guarantees of uptime when performing rolling updates on a spot pool" there would be no issue here. As it stands today, no value has been added by that limitation. I still get brief interruptions during pool rolls, but I also have to spend significantly more of my own time running manual steps.

**Comment by Keramblock** [comment]
Same here

**Comment by kaarthis** [comment]
We are assessing this right now will keep you posted.

**Comment by palma21** [comment]
According to this thread's feedback we will consider allowing spot pools to upgrade in-place by exempting them from availability guarantees (such as surge nodes, etc) that are provided on on-demand nodepools.

**Comment by paulgmiller** [comment]
The current thought here is that we'll 
1) Delete ALL Nodes in a spot agent pool from api server WITHOUT evicting pods. 
2) Pods will be terminated by their controllers if applicable but 
3) All nodes in spot pool will  reimage immediately (no surge or rolling reimage applied)

Like @palma21 I'm  curious if this meets expectations of people looking for spot pool upgrade.

**Comment by rogelioamancisidor** [comment]
It seems this problem also affects deploying spotnodes in a cluster with auto-update #2822

**Comment by paulgmiller** [comment]
@dvdvorle thanks for the feedback.  Our concern is that attempts to do rolling upgrades will have a high rate of failure. Surge may fail but we also might lose machines during the upgrade as well so its hard to know when to fail fast. 
Even if we let you specify a percent unavailable we'd still need to constantly monitor machines during the for losses.  

If we drain there is a good chance evicted pods can't actually schedule anywhere even before we start the upgade if they were previously constrained to the spot pool.

Perhaps you and @palma21 can educate me on something.  if you use preferredDuringSchedulingIgnoreDuringExecution  today and a spot pool delllocates don't you have the same problem of your more expensive pool scaling up? How do pods get kicked off back to the spot pool and why doesn't the same pattern apply for this upgrade? 

There is a decent chance reimages would actually happen faster than autoscaler spins up new machines but I don't have data on that yet.

**Comment by dvdvorle** [comment]
@paulgmiller Thank you for your explanation. I assume that monitoring the machines for losses is harder than it sounds.

When a spot pool deallocates the regular pool will scale up. We have a [overprovisioning](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-configure-overprovisioning-with-cluster-autoscaler) deployment that has a required affinity to spot nodepools that will keep trying to scale the spot node pool back up. We have currently have a basic descheduler job running to kick pods off nodes when their preferred affinity can be better satisfied by other nodes. Though I haven't found a decent descheduler yet, one that is able to emulate the scheduler as well as the real thing, see https://github.com/kubernetes-sigs/descheduler/issues/211 for example.

So kicking pods back might take half a day or so after a deallocate event, which is okay if it doesn't happen too often. Having it happen every update would feel unnecessary.

Though if what you intuit is true, about the reimaging being fast enought, than that would be probably work for us as well.

**Comment by k0eff** [comment]
During a regular node pool upgrade a new, temporary node(s) is launched (the exact number depending on surge ofc). Why not create a new, temporary _**node pool**_ with the same vm size, then cordon, rollout all deployments, drain and recreate the nodepool. If it fails because no other vms of that size are available - it will fail at step 1 and thus not affect the real node pool.

**Comment by andre-at-bede** [comment]
We're having exactly the same problem with terraform config where it's expected to have exactly the same spot node pools in place after upgrade, so destroying and recreating spot node pools with specific features (which are hold in terraform) are not feasible.
I can say it's down to client how to run the upgrade without "downtime". With spot node pools there is no guarantee there will be availability, so all nodes can be terminated anytime. So I'm thinking this is nothing to do with upgrade. If node is terminated when upgrade is running, no big deal as I'm expecting this anytime. It's down to client to provide enough capacity on standard node pools to take away workloads which run previously on spot node pools.
However iirc when it's been tested, upgrade wasn't even possible for spot node pools without any nodes in it as I've planned doing something like that:
- Have enough spot node pools with different eviction rates, so I'm confident I can move workloads around (if not I could scale out standard node pools).
- Empty spot node pool, so there is no nodes in it and make sure nothing will trigger scaling activity in this node pool (e.g. change auto to manual scaling for time of upgrade). 
- Upgrade the spot node pool and move to next one with similar steps (obviously as node pools are not destroyed, no terraform config needs to change).
It's not the best but it will be good if Azure at least provide some options which allow to work around current scenario as right know it's difficult to make this process automatic due to Azure blocking it.

**Comment by kaarthis** [comment]
Thank you for the comments our team internally have been discussing various aspects and will be going with the below approach for Spot upgrades
* Initially the idea is to go one node at a time. 
* A cordon will be issued , followed by eviction notice. 
* There wont be any surge nodes for SPOT. 
* There wont be any 'Drain' issued just eviction notice. 
* Maxunavailable for upgrades whenever it comes up in future, should apply for SPOT as well down the line (Currently there is no Maxunavailable setting)

**Comment by paulgmiller** [comment]
>>There wont be any 'Drain' issued just eviction notice.
To clarify a bit this means we won't repeat evictions till pod leave or we timeout. eviction will be one time and won't retry.

**Comment by k0eff** [comment]
Does this mean the `az aks nodepool upgrade` `--node-image-only` option will be retained for the spot nodes upgrade process? 
As for our case it will be best to just "bump" the node pool image version and leave the old nodes there. It has to be a cluster operator's decision whether to kill or keep the old nodes... As they may be killed by a descheduler policy for example or even manually or by a script. 

Ideally it has to be possible to have a scripted upgrade process with the following steps:
1. Start the upgrade process (`az aks nodepool upgrade`)
2. Have the node pool image ugpraded
3. Cordon all nodes in the node pool (either automatically by the nodepool upgrade command or manually with a for loop)
4. Restart all deployments in the node pool (effectively draining the node in a gentle way)
5. Remove all the node cordoned and gently drained nodes

**Comment by paulgmiller** [comment]
node-image-upgrade will be an option for spot node pool. But we won't leave old nodes behind (that'd be rather unexpected for most custoemrs). What you describe is more of a max surge 100% approach but that's hard with spot since you're not garunteed to get more nodes. 
Also wasn't clear what you meant by restart all deployments. Why would that be prefrable to evictions? Because it would follow maxSurge/maxunavailble on them? You could get s simliar effect with pod disruption budgets.

**Comment by pdefreitas** [comment]
The current workaround for this issue is having to delete the nodepool and recreate it with the target version which is very inconvenient for auto-upgrade scenarios like stated in #2822.

It would be nice to have this feature implemented.

**Comment by kaarthis** [comment]
This has been released and will be available for consumption shortly this week to all regions. Stay tuned for a Doc update as well.

**Comment by k0eff** [comment]
I can confirm this has been released in the most recent version of azure-cli! Thank you, AKS Team for creating this feature for us!

P.S. For anyone who stumbles upon this thread, here is a proposal for an automated cluster upgrade for control plane and node pools, which can be a base of a more extensive upgrade script.

**Comment by thejosephstevens** [comment]
This is wonderful, thank you so much for releasing this! I get to delete a whole bunch of unneeded terraform code now!
